File: 06_strings/06_09_maximum_nesting_depth_of_parentheses/maximum_nesting_depth_of_parentheses.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int maxDepth(const string& s) {
    int result = 0;
    int count = 0;
    for (char ch : s) {
        if (ch == '(') {
            count++;
            result = max(result, count);
        } else if (ch == ')') {
            count--;
        }
    }
    return result;
}

void testMaxDepth(const string& s) {
    int depth = maxDepth(s);
    cout << "Output: \"" << s << "\", Depth: " << depth << endl;
}

int main() {
    // Test cases
    testMaxDepth("(a(b)c)");           // Output: "(a(b)c)", Depth: 2
    testMaxDepth("(a(b(c)d)e)");        // Output: "(a(b(c)d)e)", Depth: 3
    testMaxDepth("()");                 // Output: "()", Depth: 1
    testMaxDepth("((()))");             // Output: "((()))", Depth: 3
    testMaxDepth("()()");               // Output: "()()", Depth: 1
    testMaxDepth("((())())");           // Output: "((())())", Depth: 4
    testMaxDepth("");                   // Output: "", Depth: 0

    return 0;
}

File: 06_strings/06_09_maximum_nesting_depth_of_parentheses/maximum_nesting_depth_of_parentheses.py
def max_depth(s: str):
    count = 0
    result = 0
    for ch in s:
        if ch == '(':
            count += 1
            result = max(count, result)
        elif ch == ')':
            count -= 1
    return result

def test_max_depth(s):
    depth = max_depth(s)
    print(f"Output: \"{s}\", Depth: {depth}")

# Test cases
test_max_depth("(a(b)c)")           # Output: "(a(b)c)", Depth: 2
test_max_depth("(a(b(c)d)e)")        # Output: "(a(b(c)d)e)", Depth: 3
test_max_depth("()")                 # Output: "()", Depth: 1
test_max_depth("((()))")             # Output: "((()))", Depth: 3
test_max_depth("()()")               # Output: "()()", Depth: 1
test_max_depth("((())())")           # Output: "((())())", Depth: 4
test_max_depth("")                   # Output: "", Depth: 0